src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_drm.c

```diff
@@ -738,154 +758,414 @@ int xocl_cleanup_mem(struct xocl_drm *drm_p)
+                xocl_grp->g_count++;
+
+                return grp_id;
+        }  
+
+        return -EINVAL;
+}
 
-		xocl_info(drm_p->ddev->dev, "  Memory Bank: %s", mem_data->m_tag);
-		xocl_info(drm_p->ddev->dev, "  Base Address:0x%llx",
-			mem_data->m_base_address);
-		xocl_info(drm_p->ddev->dev, "  Size:0x%lx", ddr_bank_size);
-		xocl_info(drm_p->ddev->dev, "  Type:%d", mem_data->m_type);
-		xocl_info(drm_p->ddev->dev, "  Used:%d", mem_data->m_used);
-	}
+int xocl_cleanup_connectivity(struct xocl_drm *drm_p)
+{
+        struct xocl_mem_map     *mem_map = NULL;
+        struct xocl_mem_group   *xocl_grp = NULL; 
+        int i = 0;
+
+        if (!drm_p->m_connect)
+                return 0;
+
+        if (drm_p->m_connect->mem_group) {
+                xocl_grp = drm_p->m_connect->mem_group;
+                /* Cleanup memory usage stats and group info for all the groups */
+                for (i = 0; i < xocl_grp->g_count; i++) {
+                        if (drm_p->mm_usage_stat && drm_p->mm_usage_stat[i]) 
+                                vfree(drm_p->mm_usage_stat[i]);
+
+                        if (xocl_grp->m_group[i])
+                                vfree(xocl_grp->m_group[i]);
+                }
+                
+                vfree(drm_p->m_connect->mem_group);
+        }
+
+        /* Cleanup memory connection info */
+        if (drm_p->m_connect->mem_map) {
+                mem_map = drm_p->m_connect->mem_map;
+                for (i = 0; i < mem_map->m_count; i++) 
+                        if (mem_map->m_map[i])
+                                vfree(mem_map->m_map[i]);
+
+                vfree(drm_p->m_connect->mem_map);
+        }
+
+        vfree(drm_p->m_connect);
+        drm_p->m_connect = NULL;
+
+        return 0;
+}
 
-	/* Initialize the used banks and their sizes */
-	/* Currently only fixed sizes are supported */
-	for (i = 0; i < topo->m_count; i++) {
-		mem_data = &topo->m_mem_data[i];
+int xocl_init_connectivity(struct xocl_drm *drm_p, struct mem_topology *topo)
+{
+        struct connectivity *connect = NULL;
+        struct xocl_mem_map *mem_map = NULL;
+        struct xocl_mem_group_info *m_group = NULL;
+        struct ip_layout *layout = NULL;
+        int group_id = -1;
+        int ip_cnt = 0;
+        int arg_cnt = 0;
+        int err = 0;
+        int i = 0;
+
+        err = XOCL_GET_IP_LAYOUT(drm_p->xdev, layout);
+        if (err)
+                return err;
+
+        if (!layout) {
+                XOCL_PUT_MEM_TOPOLOGY(drm_p->xdev);
+                return -EINVAL;
+        }
+
+        err = XOCL_GET_CONNECTIVITY(drm_p->xdev, connect);
+        if (err)
+                return err;
+
+        if (connect == NULL) { 
+                XOCL_PUT_MEM_TOPOLOGY(drm_p->xdev);
+                XOCL_PUT_CONNECTIVITY(drm_p->xdev);
+                return -ENODEV;
+        }
+
+        drm_p->m_connect = vzalloc(sizeof(struct xocl_mem_connectivity));
```

Comment:
is it possible to calculate the maximum buffer size we need to hold the maximum number of groups based on xclbin meta data? if we can, let's do one vzalloc instead of multiple ones spread around different functions. it's ok to waste some memory, if not a lot. but it'll be much easier to maintain the code and less likely to leak memory, which is a very bad bug for driver.

Reply:
Updated as per your comments. Please review.